build-system/tasks/helpers.js

```diff
@@ -150,7 +150,19 @@ async function bootstrapThirdPartyFrames(options) {
  * @param {!Object} options
  */
 async function compileCoreRuntime(options) {
-  await doBuildJs(jsBundles, 'amp.js', options);
+  async function watchFunc() {
+    const bundleComplete = await doBuildJs(jsBundles, 'amp.js', options);
+    if (options.onWatchBuild) {
+      options.onWatchBuild(bundleComplete);
+    }
+  }
+
+  if (options.watch) {
+    const debouncedRebuild = debounce(watchFunc, watchDebounceDelay);
+    fileWatch('src/**/*.js').on('change', debouncedRebuild);
```

Comment:
> You can get a head start by eliminating all the other JS targets that also live in src/:

The reason I asked the earlier question is because in the _ideal_ case, we wouldn't need to have our folder structure match our dependency structure for watching to be perfect. Our bundler is perfectly capable of handing us the full dependency tree, and we can use it to trigger rebuilds for the dependents necessary.

A very naive implementation of that would mean watching hundreds of files individually instead of efficiently rolling up to folders

Reply:
> You can get a head start by eliminating all the other JS targets that also live in src/:

The reason I asked the earlier question is because in the _ideal_ case, we wouldn't need to have our folder structure match our dependency structure for watching to be perfect. Our bundler is perfectly capable of handing us the full dependency tree, and we can use it to trigger rebuilds for the dependents necessary.

A very naive implementation of that would mean watching hundreds of files individually instead of efficiently rolling up to folders

Reply:
Fair point. There have been conversations about adopting a more modern folder structure, but that is a fair way away.